storage/replica.go

```diff
@@ -2048,7 +2079,32 @@ var enableQuiescence = envutil.EnvOrDefaultBool("COCKROACH_ENABLE_QUIESCENCE", f
 // follower which will send a MsgProp to the leader that will unquiesce the
 // leader. If the leader of a quiesced range dies, followers will not notice,
 // though any request directed to the range will eventually end up on a
-// follower which will unquiesce the follower and lead to an election.
+// follower which will unquiesce the follower and lead to an election. When a
+// follower unquiesces for a reason other than receiving a raft message or
+// proposing a raft command (for example the concurrent enqueuing of a tick),
+// it wakes the leader by sending an empty message proposal. This avoids
+// unnecessary elections due to bugs in which a follower is left unquiesced
+// while the leader is quiesced.
+//
+// Note that both the quiesce and wake-the-leader messages can be dropped or
+// reordered by the transport. The wake-the-leader message is termless so it
+// won't affect elections and, while it triggers reproprosals that won't cause
+// problems on reorderin. If the wake-the-leader message is dropped the leader
+// won't be woken and the follower will eventually call an election.
+//
+// If the quiesce message is dropped the follower which missed it will not
+// quiesce and will eventually cause an election. The quiesce message is tagged
+// with the current term and commit index. If the quiesce message is reordered
+// it will either still apply to the recipient or the recipient will have moved
+// forward and the quiesce message will fall back to being a heartbeat.
+//
+// TODO(peter): There remains a scenario in which a follower is left unquiesced
```

Comment:
I think this TODO can be removed. This scenario is already acknowledged in the previous paragraph.